### PR TITLE
kwargs in EmpiricalHistogram

### PR DESCRIPTION
**Pull request by cyborg1995**
Issue: https://github.com/JuliaEarth/GeoStats.jl/issues/153

### CODE REVIEW

**Comment by cyborg1995** [comment]
@juliohm could you please review this

File: src/statistics/data.jl [hunk 1]
```diff
@@ -50,8 +50,8 @@ Optionally, specify the variable `v` and the block side `s`.
 """
 quantile(d::AbstractData, p) = quantile(d, p, median_heuristic(d))
 
-EmpiricalHistogram(d, v::Symbol, w::WeightingMethod) = fit(Histogram, d[v], weight(d, w))
-EmpiricalHistogram(d, v::Symbol, s::Number) = EmpiricalHistogram(d, v, BlockWeighting(ntuple(i->s,ncoords(d))))
+EmpiricalHistogram(d, v::Symbol, w::WeightingMethod; kwargs...) = fit(Histogram, d[v], weight(d, w); kwargs...)
+EmpiricalHistogram(d, v::Symbol, s::Number; kwargs...) = EmpiricalHistogram(d, v, BlockWeighting(ntuple(i->s,ncoords(d))); kwargs...)
 EmpiricalHistogram(d, v::Symbol) = EmpiricalHistogram(d, v, median_heuristic(d))
 EmpiricalHistogram(d, w::WeightingMethod) = Dict(v => EmpiricalHistogram(d, v, w) for v in name.(variables(d)))
 EmpiricalHistogram(d, s::Number) = EmpiricalHistogram(d, BlockWeighting(ntuple(i->s,ncoords(d))))
```
Comment (juliohm) [inline src/statistics/data.jl:57]: Should we also forward the kwargs in these methods?

File: src/statistics/data.jl [hunk 2]
```diff
@@ -60,9 +60,11 @@ EmpiricalHistogram(d, s::Number) = EmpiricalHistogram(d, BlockWeighting(ntuple(i
     EmpiricalHistogram(sdata)
     EmpiricalHistogram(sdata, v)
     EmpiricalHistogram(sdata, v, s)
+    EmpiricalHistogram(sdata, v, s; kwargs...)
```
Comment (juliohm) [inline src/statistics/data.jl:63]: Let's try to add the kwargs to all methods, it should be very similar to what you already did for the first two methods 👍🏽

File: src/statistics/data.jl [hunk 3]
```diff
@@ -60,9 +60,11 @@ EmpiricalHistogram(d, s::Number) = EmpiricalHistogram(d, BlockWeighting(ntuple(i
...
 
 Spatial histogram of spatial data `sdata`. Optionally,
 specify the variable `v` and the block side `s`.
+Keyword arguments can also be specified.
```
Comment (juliohm) [inline src/statistics/data.jl:67]:
```suggestion
specify the variable `v`, the block side `s`, and the
keyword arguments `kwargs` for `fit(Histotogram, ...)`.
```

**Review by juliohm — CHANGES_REQUESTED** [review]
That is awesome @cyborg1995 , thanks for the PR 💯 

We just need to do the same for the other methods and add some tests to make sure that options are being passed correctly. For example, a test where the user specifies the number of bins with `bins=10` or something, where `bins` is one of the `kwargs` that can be passed to `fit(Histogram, ...)`.

**Comment by cyborg1995** [comment]
@juliohm, I added kwargs for all the methods and also the test.
But I'm not so sure about the test.

File: src/statistics/data.jl [hunk 4]
```diff
@@ -50,19 +50,24 @@ Optionally, specify the variable `v` and the block side `s`.
...
+EmpiricalHistogram(d, s::Number; kwargs...) = EmpiricalHistogram(d, BlockWeighting(ntuple(i->s,ncoords(d))); kwargs...)
 
 """
     EmpiricalHistogram(sdata)
     EmpiricalHistogram(sdata, v)
     EmpiricalHistogram(sdata, v, s)
+    EmpiricalHistogram(sdata, v, s; kwargs...)
+    EmpiricalHistogram(sdata, v; kwargs...)
+    EmpiricalHistogram(sdata, w; kwargs...)
+    EmpiricalHistogram(sdata, s; kwargs...)
```
Comment (juliohm) [inline src/statistics/data.jl:66]: I think we only need to add the `; kwargs...` piece to the existing lines, right? Can you erase the new lines created here?

**Review by juliohm — CHANGES_REQUESTED** [review]
Nice @cyborg1995 , thanks for improving the PR. I think you can mark it as ready for review at this point and GitHub will stop showing it as Draft. We can then see the results of the tests.

Regarding the test, you just need to call a EmpiricalHistogram with `bins` as you did and double check that the result indeed has this number of bins. Could you do that? :)

**Comment by juliohm** [comment]
@cyborg1995 after you are done with this PR, I have another idea of contribution where you could help already. It is also related to the EmpiricalHistogram 👍🏽

**Comment by cyborg1995** [comment]
@juliohm I would love to make more contributions!
I've called EmpiricalHistogram with nbins=10 but how do I get the number of bins from this EmpiricalHistogram to check if it's equal to what was specified (10)?

**Comment by juliohm** [comment]
@cyborg1995 according to the StatsBase.jl documentation you can for example access the `h.weights` vector which contains the heights of the bins: https://juliastats.org/StatsBase.jl/stable/empirical/#Histograms-1 This vector should have the desired number of items, in this case 10.

Please let me know if you have additional questions, happy to help :)

**Comment by cyborg1995** [comment]
@juliohm yes, that is exactly what I tried, but the length of the weight vector changes with the sdata while nbins=10 was constant. Is this possible?
```
sdata = georef((z=[1, 2, 3],))
length(EmpiricalHistogram(sdata, :z; nbins=10).weights)
```
gives `11`, and
```
sdata = georef((z=[1, 2, 3, 4],))
length(EmpiricalHistogram(sdata, :z; nbins=10).weights)
```
gives `7`

**Comment by juliohm** [comment]
That is interesting. It also happens with large vectors like `z=rand(100)` ? I thought that the nbins option would force the number of weights, perhaps this is worth asking in one of the Julia forums like Zulip?

**Comment by cyborg1995** [comment]
@juliohm Oh, I see now. This does not happen with large vectors.
This probably has something to do with empirical distributions.

File: test/data.jl [hunk 5]
```diff
@@ -47,6 +47,10 @@
     ct = DataFrame(Tables.columns(t[1:3,:]))
     @test rv == rt
     @test cv == ct
+
+    # kwargs
+    sdata = georef((z=rand(100),))
+    @test length(EmpiricalHistogram(sdata, :z; nbins=10).weights) == 10
```
Comment (juliohm) [inline test/data.jl:53]: Can you split this lines into two lines? The first computes the `h = EmpiricalHistogram(...)` and second checks the length.

**Review by juliohm — CHANGES_REQUESTED** [review]

File: test/data.jl (hunk 5 above)
Comment (cyborg1995) [reply test/data.jl:53]: Yeah sure

**Comment by cyborg1995** [comment]
The errors don't seem to be linked to this PR, could you please confirm what they are about?

File: src/statistics/data.jl [hunk 6]
```diff
@@ -50,19 +50,20 @@ Optionally, specify the variable `v` and the block side `s`.
...
-    EmpiricalHistogram(sdata, v, s)
+    EmpiricalHistogram(sdata, v; kwargs...)
+    EmpiricalHistogram(sdata, v, s; kwargs...)
 
 Spatial histogram of spatial data `sdata`. Optionally,
-specify the variable `v` and the block side `s`.
+specify the variable `v`, the block side `s`, and the
+keyword arguments `kwargs` for `fit(Histotogram, ...)`.
 """
 EmpiricalHistogram(d) = EmpiricalHistogram(d, median_heuristic(d))
```
Comment (juliohm) [inline src/statistics/data.jl:68]: I think this one is still missing the `kwargs` right?

**Comment by juliohm** [comment]
Yes, the errors should be unrelated. For some reason a dependency got updated and we didn't fix the version in the test suite. I will submit a fix after your work is merged. 👍🏽

File: src/statistics/data.jl (hunk 6 above)
Comment (cyborg1995) [reply src/statistics/data.jl:68]: Oh yes, forgot about that one.

**Review by juliohm — APPROVED** [review]

**Pull request merged**